aips/aip-103.md

```diff
@@ -0,0 +1,110 @@
+- We want such accounts to be limited to Federated Keyless Accounts, as constructing Keyless Accounts without aud checks is unsafe.  This can be mitigated by the Aptos SDK disallowing `aud`-less accounts from being used as Keyless Accounts.  The prover will also reject proof requests for Keyless providers (as of now Google and Apple).  However, in a world where 3rd party provers are permitted, we cannot prevent developers from using `aud`-less accounts as Keyless Accounts, but developers would not have any incentive construct such accounts for their users (these accounts would be accessable by any other dApp, regardless of trust).
+- The verification key will need an update, which will invalidate all existing proofs.  Additionally the prover will need to start proving with the new proving key right away after the update.  The prover has already been updated to support the proving key rotations and the SDK also supports state checks to invalidate old proofs.
+
+## Alternative solutions
+
+The alternative is to add an additional keyless public key type where the formula to compute the IdCommitment does not contain the `aud` at all.
+
+This is the advantage of explicit type safety as a completely new validation path would be implemented.  There would be no risk of such proofs being accepted for accounts that require `aud` to be present due to explicit differences in how the proof would be validated gated on the type of public key.
+
+However the drawbacks include:
+- We need to add a new keyless public key type, which may not be needed if we can leverage the existing design.  And avoiding proliferation of keyless public key types is desirable.
+- Requiring imiplementation of a new authentication path in the authenticator, which may be error prone and takes additional engineering effort.
+- Requires more complex changes to the prover as it would need to support a different public inputs hash calculation in order to differentiate between accounts with and without `aud`.  Or it would need to use a different circuit versionentirely.
+
+Thus if we can leverage the existing design, it would be preferable to do so.
+
+## Specification and Implementation Details
+
+This AIP's implementation has three parts -
+
+1. We add an additional private input into the prover.  This value will indicate whether the `aud` check is enabled.  
+- If it is enabled, the circuit will do the status quo set of verifications.  
+- If it is disabled, the circuit will use an empty `aud` value as a private input and skip checking the `aud` field.  A valid JWT still needs to be provided and the public inputs hash calculation will still implicitly include the `aud` field via the `IdCommitment` in the `KeylessPublicKey`, but it will be set to the empty value.  This means that proofs for `aud`-less accounts will be rejected if the account requires `aud` to be present, as the `IdCommitment` will be different.
```

Comment:
Not sure what you mean by "use an empty `aud` value as a private input"?

"skip checking the `aud` field" --> phrased like this, it sounds super dangerous.

In general, the description here is poor/imprecise: there are several `aud` fields flying around, so just talking about the "`aud` field" in general is ambiguous: is it the one in the JWT? Is the one in the `IdCommitment`? is it the empty `aud`?

In addition to a more precise overview of how it works, we will need to actually show the modified ZK relation, highlighting the changed parts.